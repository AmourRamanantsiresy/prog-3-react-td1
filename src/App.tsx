import { useState } from 'react';
import TodoAddForm from './common/TodoAddForm';
import TodoItemList from './common/TodoItemList';
import Todo, { TodoStatus } from './utils/models/Todo';

function App() {
  const [todos, setTodos] = useState<Todo[]>([]);

  const addTodo = (newTodo: Todo) => {
    setTodos(currentTodos => [...currentTodos, newTodo]);
  };

  const removeTodo = (todoId: number) => {
    setTodos(currentTodos => currentTodos.filter(todo => todo.id !== todoId));
  };

  const doneTodo = (todoId: number) => {
    const editTodos = (currentTodos: Todo[]) => {
      return currentTodos.map(todo => (todo.id !== todoId ? todo : { ...todo, status: TodoStatus.DONE }));
    };
    setTodos(editTodos);
  };

  return (
    <div className='container' style={{ display: 'flex', justifyContent: 'center' }}>
      <div style={{ display: 'flex', width: '50vw', justifyContent: 'space-around' }}>
        <TodoItemList
          actions={<TodoAddForm add={addTodo} />}
          todoStatus={TodoStatus.TODO}
          todos={todos}
          onDone={doneTodo}
          onRemove={removeTodo}
        />
        <TodoItemList todoStatus={TodoStatus.DONE} todos={todos} onDone={doneTodo} onRemove={removeTodo} />
      </div>
    </div>
  );
}

export default App;
