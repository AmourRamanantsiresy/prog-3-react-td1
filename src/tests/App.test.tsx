import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create, ReactTestRendererJSON } from 'react-test-renderer';
import App from '../App';
import { TI_INPUT_TODO_ADD, TR_INPUT_TODO_DELETE, TR_INPUT_TODO_DONE } from './utils/selectors';

describe('Test App', () => {
  test('Architecture', () => {
    const tree = create(<App />).toJSON() as ReactTestRendererJSON;

    expect(tree).toMatchSnapshot();
  });

  test('Application functionality', () => {
    render(<App />);

    screen.getAllByText('No task');
    screen.getByPlaceholderText('Add task');
    const inputAddSelector = screen.getByTestId(TI_INPUT_TODO_ADD);

    userEvent.type(inputAddSelector, 'New todo 1{Enter}');

    screen.getByText('New todo 1');

    const toDoneSelector = screen.getByRole(TR_INPUT_TODO_DONE);
    userEvent.click(toDoneSelector);

    userEvent.type(inputAddSelector, 'New todo 2{Enter}');
    userEvent.type(inputAddSelector, 'New todo 3{Enter}');

    screen.getByText('New todo 2');

    expect(screen.queryAllByRole(TR_INPUT_TODO_DONE).length).toBe(2);
    expect(screen.queryAllByRole(TR_INPUT_TODO_DELETE).length).toBe(1);

    const toRemoveSelector = screen.getByRole(TR_INPUT_TODO_DELETE);
    userEvent.click(toRemoveSelector);

    expect(screen.queryByRole(TR_INPUT_TODO_DELETE)).not.toBeTruthy();
  });
});
