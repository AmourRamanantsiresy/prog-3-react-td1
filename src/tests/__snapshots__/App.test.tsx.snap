// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App Architecture 1`] = `
<div
  className="container"
  style={
    Object {
      "display": "flex",
      "justifyContent": "center",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-around",
        "width": "50vw",
      }
    }
  >
    <div
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      <h1>
        TODO
      </h1>
      <div
        className="actions"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            data-testid="input-todo-add-id"
            onChange={[Function]}
            placeholder="Add task"
            type="text"
            value=""
          />
        </form>
      </div>
      <div
        className="list-container"
      >
        <p
          className="message-muted"
        >
          No task
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "padding": ".5rem",
        }
      }
    >
      <h1>
        DONE
      </h1>
      <div
        className="actions"
      />
      <div
        className="list-container"
      >
        <p
          className="message-muted"
        >
          No task
        </p>
      </div>
    </div>
  </div>
</div>
`;
